src/WebJobs.Script.WebHost/Security/KeyManagement/KubernetesSecretsRepository.cs

```diff
@@ -0,0 +1,216 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.KeyVault;
+using Microsoft.Azure.KeyVault.Models;
+using Microsoft.Azure.Services.AppAuthentication;
+using Microsoft.Azure.WebJobs.Script.IO;
+using Microsoft.Extensions.Logging;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Microsoft.Rest.Azure;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    /// <summary>
+    /// An <see cref="ISecretsRepository"/> implementation that uses the key vault as the backing store.
+    /// </summary>
+    public class KubernetesSecretsRepository : ISecretsRepository, IDisposable
+    {
+        // host.master = value
+        private const string MasterKey = "host.master";
+        // host.function.{keyName} = value
+        private const string HostFunctionKeyPrefix = "host.function.";
+        // host.systemKey.{keyName} = value
+        private const string SystemKeyPrefix = "host.systemKey.";
+        // functions.{functionName}.{keyName} = value
+        private const string FunctionKeyPrefix = "functions.";
+        private readonly IEnvironment _environment;
+        private readonly IKubernetesClient _kubernetesClient;
+        private bool _disposed;
+
+        public KubernetesSecretsRepository(IEnvironment environment, IKubernetesClient kubernetesClient)
+        {
+            _environment = environment;
+            _kubernetesClient = kubernetesClient;
+            _kubernetesClient.OnSecretChange(() =>
+            {
+                SecretsChanged?.Invoke(this, new SecretsChangedEventArgs { SecretsType = ScriptSecretsType.Host });
+                SecretsChanged?.Invoke(this, new SecretsChangedEventArgs { SecretsType = ScriptSecretsType.Function });
+            });
+        }
+
+        public event EventHandler<SecretsChangedEventArgs> SecretsChanged;
+
+        public bool IsEncryptionSupported => false;
+
+        public async Task<ScriptSecrets> ReadAsync(ScriptSecretsType type, string functionName)
+        {
+            return type == ScriptSecretsType.Host ? await ReadHostSecrets() : await ReadFunctionSecrets(functionName);
+        }
+
+        public async Task WriteAsync(ScriptSecretsType type, string functionName, ScriptSecrets secrets)
+        {
+            if (!_kubernetesClient.IsWritable)
+            {
+                throw new InvalidOperationException($"{nameof(KubernetesSecretsRepository)} is readonly when no {EnvironmentSettingNames.AzureWebJobsKubernetesSecretName} is specified.");
```

Comment:
Even though I'm throwing here if the repo is ReadOnly, update and create API calls still succeed. I'm not sure what is the right way to indicate a write error from the Repository to the SecretsManager to bubble up to the Controller?